### ISSUE TITLE
Audit all commands for goroutine leaks

### ISSUE THREAD

**Issue by Stebalien** [ipfs/go-ipfs#4414]
We're leaking goroutines on canceled commands all over the place.

To do this audit, I recommend looking at all channel writes and I mean **all**.

A quick grep indicates that the following files *may* have issues (should, at least, be looked at):

* [x] `core/commands/dht.go` -- #4413
* [x] `core/commands/repo.go` -- #4413
* [ ] `core/commands/add.go`
* [ ] `core/commands/dag/dag.go`
* [ ] `core/commands/files/files.go`
* [ ] `core/commands/commands.go`
* [ ] `core/commands/filestore.go`
* [ ] `core/commands/object/object.go`
* [ ] `core/commands/pin.go`
* [ ] `core/commands/ping.go`
* [ ] `core/commands/unixfs/unixfs.go`
* [ ] `core/commands/refs.go`

Basically, anytime we write to a channel that may block (i.e., doesn't have a buffer at least as large what we're writing), we should select on the channel write and `<-req.Context().Done()`.

**Comment by keks** [comment]
@Stebalien Regarding your question in #4413:
> Should we be calling res.SetError(req.Context().Err(), cmdkit.ErrNormal) when the context is canceled or can we just bail (what I'm doing here).

When we cancel, the error should either be correctly handled or the command should return an error indicating the command was cancelled. Maybe even specify `"go-ipfs-cmds".Canceled` of type `cmdkit.Error`, analogue to `"context".Canceled`. If this doesn't happen, we should add something like
```go
select {
case <-ctx.Done():
  res.SetError(ctx.Err(), cmdkit.ErrNormal)
case default:
}
```
at the end of the respective commands. Does that sound reasonable to you?

**Comment by Stebalien** [comment]
> Does that sound reasonable to you?

It's reasonable, just verbose and painful; in some cases, we realize that the context has been canceled in multiple places concurrently (thread safety issues). If the context has been canceled/expired, I assume the client isn't waiting for a response anyways. If they are, can the commands package not handle this?

**Comment by Stebalien** [comment]
Note: We *should not* be closing the channel on context cancel. Otherwise, the other side might think that we're done sending the response.

**Comment by forstmeier** [comment]
Hi, all! Is this issue still active? I'd like to get involved with some entry-level work on the project to get my hands dirty and this issue seemed like it still needed some work done. :smile:

**Comment by alecbrick** [comment]
@forstmeier I'm interested in getting into IPFS as well, and I'm also interested in working on this issue. It doesn't seem super active, but if you're interested, we might be able to jointly look into this. Some of the files seem clean, but I'm not super experienced at Go, and I'm not great at intuitively knowing when a goroutine might hang.

**Comment by forstmeier** [comment]
Hey, @alecbrick, I saw that you've submitted fixes for dag.go and pin.go - do you want me to look through refs.go?

**Comment by kjzz** [comment]
@Stebalien there are one goroutine leaks in filestore.go.So please help me review pr #5427 ,Thx a lot

**Comment by tarekbadrshalaan** [comment]
Hi @Stebalien,
I tried to contribute in this issue, found four file still remaining. 

- `core/commands/commands.go`
[go-ipfs/blob/master/core/commands/commands.go](https://github.com/ipfs/go-ipfs/blob/master/core/commands/commands.go#L141-L152)
- `core/commands/object/object.go`
[go-ipfs/blob/master/core/commands/object/object.go](https://github.com/ipfs/go-ipfs/blob/master/core/commands/object/object.go#L598-L609)
- `core/commands/unixfs/unixfs.go` 
[go-ipfs/blob/master/core/commands/unixfs/unixfs.go](https://github.com/ipfs/go-ipfs/blob/master/core/commands/unixfs/unixfs.go#L31-L42)

the code in those file is the same. 
As you can see no context need to check if it's done.
The enhacemnet we can do is the code is duplicate, we can make it once in "commands.go".

- The last one is  `core/commands/add.go` 
[go-ipfs/blob/master/core/commands/add.go](https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L270-L360)
I don't think channel with context need more maintain.

so please tell me if you need more help in this issue or any other, I am looking for contribute with you.

**Comment by tarekbadrshalaan** [comment]
Hi @Stebalien,
I tried to contribute in this issue, found four file still remaining.

- `core/commands/commands.go`
https://github.com/ipfs/go-ipfs/blob/dcecab6854e97c5775ae57abc1bc8429b6e122da/core/commands/commands.go#L141-L152
- `core/commands/object/object.go`
https://github.com/ipfs/go-ipfs/blob/dcecab6854e97c5775ae57abc1bc8429b6e122da/core/commands/object/object.go#L598-L609
- `core/commands/unixfs/unixfs.go` 
https://github.com/ipfs/go-ipfs/blob/dcecab6854e97c5775ae57abc1bc8429b6e122da/core/commands/unixfs/unixfs.go#L31-L42

the code in those file is the same.
As you can see no context need to check if it's done.
The enhacemnet we can do is the code is duplicate, we can make it once in "commands.go".

--------------
- The last one is  `core/commands/add.go` 
every channel with context are mintained.
https://github.com/ipfs/go-ipfs/blob/dcecab6854e97c5775ae57abc1bc8429b6e122da/core/commands/add.go#L270-L360


so please tell me if you need more help in this issue or any other, I am looking for contribute with you.

**Comment by Stebalien** [comment]
The `unwrapOutput` cases are actually fine. Those are only called on the client and we close the channel when the request finishes or is canceled.

I wouldn't bother deduplicating them either. Instead, we should focus on switching to the new Commands API that doesn't need that helper (you'll notice that some commands don't use `SetOutput` or `Output`, those are the ones using the new API).

---

That last `add` one looks fine. We're waiting on the context at the bottom. Am I missing something?

**Comment by tarekbadrshalaan** [comment]
Hi @Stebalien,
- No you don't miss any point
- if you need help in this point tell me more, I need to help.

> we should focus on switching to the new Commands API that doesn't need that helper (you'll notice that some commands don't use `SetOutput` or `Output`, those are the ones using the new API).

**Comment by Stebalien** [comment]
It looks like there are some cases in core/commands/pin.go and one in core/commands/refs.go (WriteEdge "checks" the context but the check is racy).

**Comment by Stebalien** [comment]
This was fixed.
